### ISSUE TITLE
COSX Grid Issues with 146 Spherical Grid Points + ROBUST Grid Pruning Scheme

### ISSUE THREAD

**Issue by davpoolechem** [psi4/psi4#2890]
Hi, all! I mentioned this issue yesterday at the Psi4 Dev meeting, and I wanted to post the issue here, with more collected information.

The crux of the issue, as I have determined, is that the COSX code, when using the ROBUST grid pruning scheme in conjunction with 146 spherical points on either of its grids (initial or final), simply breaks.

Consider the following input:
```
molecule mol {
    0 1
    O
    H 1 0.96
    H 1 0.96 2 104.5
    symmetry c1
    no_reorient
    no_com
}

set {
    scf_type cosx
    df_scf_guess false
    basis aug-cc-pVDZ
    e_convergence 1.0e-10
    incfock true
    incfock_full_fock_every 4
    ints_tolerance 1.0e-12
    cosx_spherical_points_final 146
    cosx_radial_points_final 50
    cosx_pruning_scheme robust
}

energy = energy('scf')
```

A simple expansion of the final COSX grid compared to default. This calculation breaks upon entering the final SCF iteration (where the tight grid is used) with the following error:
```
!----------------------------------------------------------------------------------!
!                                                                                  !
!  ADIIS minimization failed. File a bug, and include your entire input and output !
!     files.                                                                       !
!                                                                                  !
!----------------------------------------------------------------------------------!
```

The following input, an expansion of both the initial and final grids, also fails:
```
molecule mol {
    0 1
    O
    H 1 0.96
    H 1 0.96 2 104.5
    symmetry c1
    no_reorient
    no_com
}

set {
    scf_type cosx
    df_scf_guess false
    basis aug-cc-pVDZ
    e_convergence 1.0e-10
    incfock true
    incfock_full_fock_every 4
    ints_tolerance 1.0e-12
    cosx_spherical_points_initial 146 
    cosx_radial_point_initial 50 
    cosx_spherical_points_final 302
    cosx_radial_points_final 75 
    cosx_pruning_scheme robust
}

energy = energy('scf')
```

This one fails in the SAD-guess iteration, with the following error:
```
!----------------------------------------------------------------------------------!
!                                                                                  !
! Fatal Error: Matrix::diagonalize: DSYEV failed!                                  !
! Error occurred in file:                                                          !
!     /theoryfs2/ds/dpoole34/Documents/Codes/psi4/psi4/src/psi4/libmints/matrix.cc !
!     on line: 1715                                                                !
! The most recent 5 function calls were:                                           !
! psi::PsiException::PsiException(std::__cxx11::basic_string<char,                 !
!     std::char_traits<char>, std::allocator<char> >, char const*, int)            !
! psi::Matrix::diagonalize(psi::Matrix&, psi::Vector&, psi::diagonalize_order)     !
!                                                                                  !
!----------------------------------------------------------------------------------!
```

Various things can be switched, with equally varying effects:
* Switching the COSX_PRUNING_SCHEME from ROBUST to TREUTLER  fixes these calculations.
* Switching from 146 spherical points to either 110 or 170 fixes these calculations.
* Switching from 50 radial points to either 55 or 45 does _not_ fix these calculations.
* Switching from water to benzene does _not_ fix these calculations.

Finally, from my observation, the problem seems to be localized specifically to COSX. Consider the following input:
```
molecule mol {
    0 1
    O
    H 1 0.96
    H 1 0.96 2 104.5
    symmetry c1
    no_reorient
    no_com
}

set {
    scf_type direct
    df_scf_guess false
    basis aug-cc-pVDZ
    e_convergence 1.0e-10
    incfock true
    incfock_full_fock_every 4
    ints_tolerance 1.0e-12
    dft_spherical_points 146
    dft_radial_points 50
    dft_vv10_spherical_points 146
    dft_vv10_radial_points 50
    dft_pruning_scheme robust
    mbis_spherical_points 146
    mbis_radial_points 50
    mbis_pruning_scheme robust
}

energy, wfn = energy('b3lyp-nl', return_wfn=True)
oeprop(wfn, "MBIS_CHARGES", "MBIS_VOLUME_RATIOS")
```

This calculation uses other components of Psi4 (DFT, VV10, MBIS) which enable specification of grid points and/or pruning schemes. This calculation, using 146 spherical points and a ROBUST pruning scheme where allowable, runs fine. 

Given that this issue is localized to COSX seemingly, I will further explore this and see what I find.

**Comment by JonathonMisiewicz** [comment]
Obligatory @hokru ping.

**Comment by hokru** [comment]
Thanks I will look into it.

**Comment by hokru** [comment]
**General comments**: The `robust` scheme uses the following orders in the regions: [7, nominal-7, nominal, nominal] Meaning only the inner 50% are pruned.  And things will break i think when the order is 7 -> 86 points or smaller.
Also for `Treutler` it does not make much sense to go below 11 -> 194 points because the order of the core region is fixed at that and you don't want to have the outer regions with less points than the inner.
I didn't take into account people would want to do pruning with small grids to be honest and even if things do not "break" there will certainly an unwanted imbalance. This needs to be at least checked by the code in the future.
As for pruning schemes for small grids, one could come up with a new scheme or even let the user input an array with orders to make grid pruning fully manually adjustable.

**On topic**: I don't really understand why this should happen currently.

**Comment by hokru** [comment]
Oh sorry i looked in the wrong column for the order<->points reference. Follow the documentation for the number of points

**Comment by hokru** [comment]
I have a working psi4 version again and will do some testing.

Using `print 5` I see the Fock matrix becomes `nan` at the end which is definitely not good :-)

I don't think ADIIS should turn on for the final iteration.

**Comment by davpoolechem** [comment]
After some deep digging, the source of the `nan`s has been identified!

In the function `compute_numeric_overlap` in DFJCOSK.cc, we have this block of code:
```
for (size_t p = 0; p < npoints_block; p++) {
    for (size_t k = 0; k < nbf_block; k++) {
        X_blockp[p][k] = point_values->get(p, k) * std::sqrt(w[p]);
    }
}
```

which is meant to compute a part of the X matrix, as defined in Eq. 2 of [this paper.](https://doi.org/10.1063/1.3646921) Problem is, with a ROBUST grid pruning scheme and 146 spherical points, some of the weights w seem to evaluate as negative. The resulting square-rooting of these weights during computing X_blockp, then, is what leads to the `nan`s that were observed.

This sort of leaves the question - how do we handle this? One obvious start is to simply have COSX throw an exception if a grid with negative weights is used.

**Comment by JonathonMisiewicz** [comment]
If you and Holger are convinced that negative grid weights _are not a bug in Psi's implementation of whatever algorithm generated those grid weights_, then indeed, throw an exception. I'd do this in the function that generates grid weights.

**Comment by susilehtola** [comment]
Negative grid weights are definitely a bug, since Becke weights are always positive. I have heard that some angular Lebedev grids reported in the literature have negative weights, but this should only affect the biggest grids.

**Comment by davpoolechem** [comment]
Hi all! I did a little more exploration into this. It seems that, as I suspected, there are other grid configurations that can crash COSX. From the testing I did, here are some different combinations of pruning scheme and spherical points that crash the COSX code:

ROBUST: 74, 146, 230, 266
TREUTLER: 74, 230, 266
NONE: 74, 230, 266 

Given that I set these spherical points for the initial COSX grid and that the resulting exception was a DSYEV failure, as in the initial posting of this issue, my guess is that these grid configurations lead to negative weights, as well.

It seems that, overall, the observed crashing is more dependent on the number of spherical points in the grid than the pruning scheme itself, although the latter has some impact.

**Comment by TiborGY** [comment]
One thing that I would like to note is that prior to https://github.com/psi4/psi4/pull/2686 this DSYEV failure would not have been caught, and Psi4 would have carried on with whatever the failed diagonalization returned.

**Comment by davpoolechem** [comment]
Based on the discussion we had today in the Psi4 dev meeting, it has been decided that the best way to fully solve this issue is simply to have COSX error out when a grid with negative weights is encountered. This is what https://github.com/psi4/psi4/pull/2906 implements, so once that PR is merged into Psi4, this issue can be closed.

**Comment by zachglick** [comment]
Two questions:
 - Have you checked the magnitude of the negative grid weights? Are they near-zero, within some small tolerance? That could be justification for zeroing them out?
 - Why not implement my above suggestion of forming two separate `X` matrices? This would be simple to do

**Comment by hokru** [comment]
> * Have you checked the magnitude of the negative grid weights? Are they near-zero, within some small tolerance? That could be justification for zeroing them out

They are of proper magnitude. Grid points with tiny weights are also already screened out during construction.

**Comment by davpoolechem** [comment]
> Two questions:
> 
> * Have you checked the magnitude of the negative grid weights? Are they near-zero, within some small tolerance? That could be justification for zeroing them out
> * Why not implement my above suggestion of forming two separate `X` matrices? This would be simple to do

@hokru answered the first part; and while I couldn't say what a "proper magnitude" is, my observation was that some of the negative weights were large enough in magnitude to where zeroing them out wouldn't be feasible.

For your second part, it's just what was settled on at the Psi4 dev meeting, moreso than anything. On further reflection, though, I can work on your suggestion today, because you're right, it wouldn't be too hard to do what you said.

**Comment by JonathonMisiewicz** [comment]
_If_ we go with Zach's suggestion, then I insist that the code emit a warning when there are negative grid weights with COSX. We're in a case where the original COSX procedure is undefined, and the user should know that, whether our fix is intuitive or not. (I don't know COSX theory well enough to comment on that.)

**Comment by davpoolechem** [comment]
> _If_ we go with Zach's suggestion, then I insist that the code emit a warning when there are negative grid weights with COSX. We're in a case where the original COSX procedure is undefined, and the user should know that, whether our fix is intuitive or not. (I don't know COSX theory well enough to comment on that.)

I would agree with this.

**Comment by susilehtola** [comment]
You can't set negative weights to zero since that will change the results. The real fix would be to use grids with non-negative weights. Using a product grid instead of Lebedev would be one option. For now I would just disable the grids with negative weights as in done in many other codes.

Switching to the use of the asymmetric form which avoids the square roots of the weights may be an option but it might also increase memory costs by a factor of 2.

**Comment by susilehtola** [comment]
The best solution would probably be to carry the minus sign for negative weights into the definition of the $A_{\nu \tau}({\bf r}_g)$, as then one has symmetric expressions with respect to $ X_{\kappa g} = \sqrt{|w_g|} \kappa ({\bf r}_g) $

**Comment by davpoolechem** [comment]
Okay, so I put up a PR with a COSX reformulation based on the suggestions @zachglick and @susilehtola made in this comment chain! Feel free to take a look there.
